ant-node/src/put_validation.rs

```diff
@@ -706,13 +706,38 @@ impl Node {
                 }
             };
 
-            if let Some(relayer_peer_id) = ant_networking::get_relay_peer_if_any(peer_multiaddr) {
-                let relayer_rewards_addr =
-                    self.network().get_rewards_address(relayer_peer_id).await?;
-                if &Some(relayer_rewards_addr) != relayer_addr {
-                    warn!("Payment quote has wrong relayer rewards address for record {pretty_key}, expected {relayer_rewards_addr:?}, got {relayer_addr:?}");
+            // See if there are any relayer peers in the peer addresses.
+            // Can be empty.
+            let relay_peers: Vec<_> = peer_multiaddrs
+                .into_iter()
+                .filter_map(ant_networking::get_relay_peer_if_any)
+                .collect();
+
+            // todo: Optimize this by including the full rewards address proof in the payment.
+            // Obtain reward addresses for relay peers.
+            let mut relay_peers_reward_addresses = vec![];
+            for relay_peer in relay_peers {
+                let relay_peer_reward_address =
+                    self.network().get_rewards_address(relay_peer).await?;
+                relay_peers_reward_addresses.push(relay_peer_reward_address);
+            }
+
+            // We expect a relayer peer to be paid here.
+            if !relay_peers_reward_addresses.is_empty() {
+                if let Some(relayer_rewards_addr) = relayer_addr {
+                    // Check if the paid relayer address was valid.
+                    if !relay_peers_reward_addresses.contains(relayer_rewards_addr) {
+                        warn!("Payment quote has wrong relayer rewards address for record {pretty_key}, expected {relayer_rewards_addr:?}, got {relayer_addr:?}");
+                        return Err(Error::InvalidRequest(format!(
+                            "Payment quote has wrong relayer rewards address for record {pretty_key}, expected {relayer_rewards_addr:?}, got {relayer_addr:?}"
+                        )));
```

Comment:
Does that mean we accept it if only one of the relayers is paid?

Reply:
Hi @grumbach ,

Yes, eventually I think we should return more connection info with every (get quote) query response, like the actual peer address that served the response. The client should then favour paying the relayer server that actually handled the get quote request (from reading the peer address in the new connection info). 

At the moment the client favours paying the latest added peer address.